extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTools.java

```diff
@@ -283,58 +385,144 @@ private static File prepareDestinationFileLocation(@NotNull File destination) {
-     * @param tableDefinition The common schema for all the tables to write
+     * @param definition The common schema for all the tables to write
      * @param writeInstructions Write instructions for customizations while writing
-     * @param destinations The destinations paths. Any non existing directories in the paths provided are created. If
+     * @param destinations The destinations paths. Any non-existing directories in the paths provided are created. If
      *        there is an error any intermediate directories previously created are removed; note this makes this method
      *        unsafe for concurrent use
      * @param groupingColumns List of columns the tables are grouped by (the write operation will store the grouping
      *        info)
      */
     public static void writeParquetTables(@NotNull final Table[] sources,
-            @NotNull final TableDefinition tableDefinition,
+            @NotNull final TableDefinition definition,
             @NotNull final ParquetInstructions writeInstructions,
             @NotNull final File[] destinations,
             @NotNull final String[] groupingColumns) {
         Require.eq(sources.length, "sources.length", destinations.length, "destinations.length");
-        if (tableDefinition.numColumns() == 0) {
+        if (definition.numColumns() == 0) {
             throw new TableDataException("Cannot write a parquet table with zero columns");
         }
-        final File[] absoluteDestinations =
+        // TODO Is there a better way to do this?
+        for (final File destFile : destinations) {
+            deleteBackupFile(destFile);
+        }
+        // Write tables to a temporary shadow file paths in the same directory to prevent overwriting any existing files
+        final File[] shadowDestFiles =
                 Arrays.stream(destinations)
-                        .map(File::getAbsoluteFile)
+                        .map(ParquetTools::getShadowFile)
                         .toArray(File[]::new);
         final File[] firstCreatedDirs =
-                Arrays.stream(absoluteDestinations)
+                Arrays.stream(shadowDestFiles)
                         .map(ParquetTools::prepareDestinationFileLocation)
                         .toArray(File[]::new);
-        for (int i = 0; i < sources.length; i++) {
-            final Table source = sources[i];
-            try {
-                writeParquetTableImpl(source, tableDefinition, writeInstructions, destinations[i], groupingColumns);
-            } catch (RuntimeException e) {
-                for (final File destination : destinations) {
-                    // noinspection ResultOfMethodCallIgnored
-                    destination.delete();
+        ArrayList<File> cleanupFiles = new ArrayList<>();
+        ArrayList<File> rollbackFiles = new ArrayList<>();
+        try {
+            ArrayList<Map<String, ParquetTableWriter.GroupingColumnWritingInfo>> groupingColumnWritingInfoMaps = null;
```

Comment:
I think you probably want an "inner write" function that accumulates rollback, shadow, and cleanup files in a set of lists that are passed in and mutated. That way, the only thing that changes between single and multiple write it the looping.
_OR_
We can make single table write just call multi-table write.

I don't think we should duplicate this much code.